File: python/test_data.py
"""
Loads Pair the Numbers test cases from ../test_data/{simple,medium,hard}/*.txt.

File format:

    name=<case name>
    target=<int>
    expected_i=<int>
    expected_j=<int>
    numbers=
    <value 0>
    <value 1>
    ...
"""

import os
from dataclasses import dataclass
from typing import List, Tuple

TEST_DATA_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "test_data")


@dataclass
class TestCase:
    name: str
    numbers: List[int]
    target: int
    expected: Tuple[int, int]


def load_case(path: str) -> TestCase:
    with open(path) as f:
        lines = f.read().splitlines()

    name = None
    target = None
    expected_i = None
    expected_j = None
    numbers_start = None

    for i, line in enumerate(lines):
        if line == "numbers=":
            numbers_start = i + 1
            break
        key, _, value = line.partition("=")
        if key == "name":
            name = value
        elif key == "target":
            target = int(value)
        elif key == "expected_i":
            expected_i = int(value)
        elif key == "expected_j":
            expected_j = int(value)

    numbers = [int(line) for line in lines[numbers_start:]]
    return TestCase(name=name, numbers=numbers, target=target, expected=(expected_i, expected_j))


def load_tier(tier: str) -> List[TestCase]:
    tier_dir = os.path.join(TEST_DATA_DIR, tier)
    cases = []
    for filename in sorted(os.listdir(tier_dir)):
        if filename.endswith(".txt"):
            cases.append(load_case(os.path.join(tier_dir, filename)))
    return cases

File: java/src/TestData.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Loads Pair the Numbers test cases from ../test_data/{simple,medium,hard}/*.txt
 * (relative to the java/ directory -- the scripts in java/scripts/ always run
 * from there).
 *
 * File format:
 *
 *     name=&lt;case name&gt;
 *     target=&lt;long&gt;
 *     expected_i=&lt;int&gt;
 *     expected_j=&lt;int&gt;
 *     numbers=
 *     &lt;value 0&gt;
 *     &lt;value 1&gt;
 *     ...
 */
public class TestData {

    static final String TEST_DATA_DIR = "../test_data";

    static class TestCase {
        final String name;
        final int[] numbers;
        final long target;
        final int expectedI;
        final int expectedJ;

        TestCase(String name, int[] numbers, long target, int expectedI, int expectedJ) {
            this.name = name;
            this.numbers = numbers;
            this.target = target;
            this.expectedI = expectedI;
            this.expectedJ = expectedJ;
        }
    }

    static TestCase loadCase(String path) throws IOException {
        List<String> lines = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader(path))) {
            String line;
            while ((line = reader.readLine()) != null) {
                lines.add(line);
            }
        }

        String name = null;
        long target = 0;
        int expectedI = 0;
        int expectedJ = 0;
        int numbersStart = -1;

        for (int i = 0; i < lines.size(); i++) {
            String line = lines.get(i);
            if (line.equals("numbers=")) {
                numbersStart = i + 1;
                break;
            }
            int eq = line.indexOf('=');
            String key = line.substring(0, eq);
            String value = line.substring(eq + 1);
            if (key.equals("name")) {
                name = value;
            } else if (key.equals("target")) {
                target = Long.parseLong(value);
            } else if (key.equals("expected_i")) {
                expectedI = Integer.parseInt(value);
            } else if (key.equals("expected_j")) {
                expectedJ = Integer.parseInt(value);
            }
        }

        List<String> numberLines = lines.subList(numbersStart, lines.size());
        int[] numbers = new int[numberLines.size()];
        for (int i = 0; i < numbers.length; i++) {
            numbers[i] = Integer.parseInt(numberLines.get(i));
        }
        return new TestCase(name, numbers, target, expectedI, expectedJ);
    }

    static List<TestCase> loadTier(String tier) throws IOException {
        File dir = new File(TEST_DATA_DIR, tier);
        File[] files = dir.listFiles((d, name) -> name.endsWith(".txt"));
        List<TestCase> cases = new ArrayList<>();
        if (files != null) {
            Arrays.sort(files);
            for (File f : files) {
                cases.add(loadCase(f.getPath()));
            }
        }
        return cases;
    }
}
